sql/core/src/main/java/org/apache/spark/sql/execution/datasources/v2/BroadcastedHadoopConf.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.datasources.v2;
+
+
+import org.apache.hadoop.conf.Configuration;
+
+import org.apache.spark.SparkContext;
+import org.apache.spark.annotation.DeveloperApi;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+import org.apache.spark.util.SerializableConfiguration;
+
+import java.util.Map;
+
+import scala.reflect.ClassTag;
+/**
+ * A helper interface to serialize and broadcast the Hadoop configuration for readers.
+ */
+@DeveloperApi
+public interface BroadcastedHadoopConf {
+  SparkSession sparkSession();
+  CaseInsensitiveStringMap options();
```

Comment:
I don't think that the data source should be responsible for providing a SparkSession. Spark has a session and configures the data source.

I think this interface should allow Spark to pass a broadcasted Hadoop Configuration to the source instead.

Reply:
I don't think I understand what you mean here, could you try and rephrase it? Sorry.

Reply:
To clarify: do you mean you want us to use reflection for something implementing this interface inside of the DSv2 code path? Or do you mean something else?

Reply:
Data sources aren't passed a Spark session, so it doesn't make sense that this interface requires the source to provide one. If we think sources need access to a Spark session, then we should add an interface that sets the session directly. (I know the built-in sources have one, but that's because those are v1 sources that have been slightly modified to fit into v2.)

The benefit of having an interface like this is to avoid needing the Spark session. That's  currently done by accessing the Spark session from the environment (e.g. `SparkSession.builder.getOrCreate()`) and we want to avoid making sources do that.

That's why I think the way this works should be to set a broadcasted configuration on a `Table` implementation that extends this interface using a method like `setBroadcastedConf`.

I'm on ASF slack if you'd like to talk about this more directly.

Reply:
Ok so we did a quick chat. It’s possible that this and one other small use case are the only reasons the sources currently have a Spark Session inside of them and the design goal
of DSv2 was to get rid of that. So I’ll see if this can be avoided, but if it’s more than broadcasting the Hadoop conf and case sensitive options then I’ll switch the API to just pass in a Spark session and we can stick with this API. Does that sound like a reasonable summary @rdblue?